### PR TITLE
New physics overrides

### LINKED ISSUE #11460: Add more movement physics attributes to modify with set_physics_override
## The status quo

Currently, with `player:set_physics_override`, you can modify three player physics attributes: Walk speed, jump strength and gravity. But I realized those are not the only things that affect movement. Looking at the various `movement_` settings in `minetest.conf`, there's more you could modify.

## The idea

Here are some new values I propose to be added to `set_physics_override`:

* `speed_climb`: Climbing speed multiplier (based on `movement_speed_climb` setting)
* `liquid_fluidity`: Liquid fluidity multiplier (based on `movement_liquid_fluidity` setting)
* `liquid_fluidity_smooth`: Multiplier for `movement_liquid_fluidity_smooth` setting
* `liquid_sink`: Liquid sinking speed multiplier (based on `movement_liquid_sink`)

Each of these settings functions like `speed`, `jump` and `gravity` right now. They are a multiplier that multiplies the set `movement_` settings for the given player. Each of them defaults to `1.0` for full backwards-compability.

There other `movement_*` settings. Not sure if they should be added to the suggestion as well. Those are a bit more esoteric and includes stuff like sneak speed and Fast Mode speed.

## Example

The API would be basically an extension of `set_physics_override`. It would look like this:

```
-- Apply a +10% climb speed boost to player
player:set_physics_override({speed_climb=1.1})
```

## Use cases

This can be used to give players certain movement bonuses or penalties for ladders and liquids, like status effects based on gear, potions, or whatever.

For example, you could give players unlockable abilities to boost movement a little:

* Fish Magic: Your liquid movement resistance is halved (`liquid_fluidity`), and you won't sink automatically in liquids (`liquid_sink`)
* Water Walker: If you fall on a liquid, you quickly come to a halt (`liquid_fluidity_smooth`)
* Master Climber: You can climb nodes faster (`speed_climb`)

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #11460 which is supported by core dev.

This PR adds the following new fields to `set_physics_override`:

* `speed_climb`: Climbing speed multiplier (based on `movement_speed_climb` setting)
* `speed_crouch`: Sneak speed multiplier (based on `movement_speed_crouch` setting)
* `liquid_fluidity`: Liquid fluidity multiplier (based on `movement_liquid_fluidity` setting)
* `liquid_fluidity_smooth`: Multiplier for `movement_liquid_fluidity_smooth` setting
* `liquid_sink`: Liquid sinking speed multiplier (based on `movement_liquid_sink`)
* `acceleration_default`: Multiplier for `movement_acceleration_default`
* `acceleration_air`: Multiplier for `movement_acceleration_air`

Those work like the existing `gravity`, `jump` and `speed`. With this update you will be able to modify certain parts of the player movement physics on a per-player basis.

To learn how it works, see the documentation updates in this PR.

## Special notes
Note about all `liquid_*` modifiers: Only nodes with `liquid_move_physics=true` are considered to be liquids for these modifiers. Don't worry, this field is true by default for nodes with `liquidtype~="none"`, i.e. most "traditional" liquid nodes.

Note about `liquid_fluidity`: This is based on `movement_liquid_fluidity` setting. But this setting currently behaves a bit buggy for very small values and is completely broke for value 0. Thus, the documentation states that `liquid_fluidity` values <1 are taboo. The reason for this is #11464. Note that most values for `liquid_fluidity` that are <1 are probably still fine and only those close to 0 are problematic, but I'm not taking any chances here, since speed in liquids depends on various other factors. As long you stay at 1 or larger for `liquid_fluidity`, this attribute is safe, according to my tests. Even `math.huge` works.

Note about `speed_climb`: As stated in the documentation, the actual climb speed is `speed` times `speed_climb`, not `speed_climb` alone. This is unfortunate, and obviously it would be better if those two were completely separate, but I can't change that. Note that `speed` is of course much older than `speed_climb` and yes, its current behavior (before the PR) also modifies the climbing speed, so that behavior has to stay for compability. Also, with some simple math, modders should be able to deal with this. 

## To do

Ready for Review.

## How to test

* In DevTest with `luacmd` mod, call `me:set_physics_override(table_of_physics_you_want_to_test)`
* Check if it does what you expect to / what the documentation promises
* For liquid, use the "move resistance" nodes in DevTest (those with the letter R on it). The `liquid_*` modifiers should affect the "liquidlike" move resistance nodes (as well as "classic" liquids like water and lava). The other move resistance nodes (non-liquidlike) should NOT be affected by the `liquid_*` modifiers.

## What I did NOT do

* Test if anti-cheat doesn't have any false-positives. Probably only relevant for `climb_speed` as it's the only one that can make you faster than the base speed. And I am guessing `climb_speed` might actually already be covered by the lax Y speed check due to bouncy nodes (`PlayerSAO::checkMovementCheat`)
* Compability with old clients/servers (I think 5.5.0 will already be incompatible with 5.4.0 anyway, so I didn't bother)
* Change the core player physics logic in any way. All I did is expose some modifiers for `movement_*` settings. If you don't ever all `set_physics_override`, everything should be the same as before

## Use cases, justification, rationale, etc.

See #11460.

### CODE REVIEW

**Comment by sfan5** [comment]
While you're at it please add a way to set the crouch movement speed

**Comment by Wuzzy2** [comment]
Oh right, I just forgot about this …

**Comment by Wuzzy2** [comment]
`speed_crouch` has been added and I did a quick successful test.

File: src/client/content_cao.cpp
Comment (sfan5) [inline src/client/content_cao.cpp:1745]: These will probably read 0.0 when values are missing and break stuff on older servers.

**Comment by Wuzzy2** [comment]
Unfortunately, I don't see a way to add a fallback mechanism right now. :-/

**Comment by sfan5** [comment]
```cpp
    float a = readF32(is);
    float b = readF32(is);
    float c = readF32(is);
    if (is.eof()) {
        a = b = c = 1.0f;
    }
```
?

**Comment by Wuzzy2** [comment]
Your guess was correct. I did a quick test and sure enough, connecting as 5.5.0-dev to 5.4.1 means your speed is set to 0, making it unplayable. Nice catch!
I have added the neccessary fallback and just to be sure it actually works, I tested with 5.5.0-dev client at 5.5.0-dev server, and 5.5.0-dev client to 5.4.1 server. Both work as expected now.

I did a rebase, the previous commits are unchanged. Only the fallback check commit is new.

**Comment by Wuzzy2** [comment]
Due to the recent merge of #10810, I had to rebase/merge this PR once again, to make it compatible with master again.
Thanks to #10810, `move_resistance` and `liquid_viscosity` are now two separate node properties. This required me to update this PR.

The update has the following implications

* *The `liquid_` modifiers only affect nodes with `liquid_move_resistance` (field from aforementioned PR)*
* Nodes with `move_resistance > 0` and `liquid_move_physics == true` WILL be affected by the `liquid_*` modifiers
* Nodes with `move_resistance > 0` and `liquid_move_physics == false` WILL NOT be affected by the `liquid_*` modifiers. The speed change is constant. This is intentional.
* All changes should still be backwards-compatible

Don't worry about `liquid_move_resistance` missing in most node definitions, it is `true` by default if `liquidtype ~= "none"`.

**Comment by Wuzzy2** [comment]
I've added `movement_speed_climb` and its override to anticheat, so that anticheat doesn't stop players with a (legit) high climbing speed.

**Comment by Wuzzy2** [comment]
Bump.

File: src/client/clientenvironment.cpp
Comment (sfan5) [inline src/client/clientenvironment.cpp:230]:
```suggestion
				if (in_liquid_stable)
					dl = MYMIN(dl, fluidity_smooth);
```

File: src/client/content_cao.cpp
Comment (sfan5) [inline src/client/content_cao.cpp:1771]: At this point this should really be a struct but oh well...

File: src/server/player_sao.cpp
Comment (sfan5) [inline src/server/player_sao.cpp:620]: This should be below the `*= 2.0` line

**Review by sfan5 — COMMENTED** [review]
Mostly works other than that, see comment.

**Comment by sfan5** [comment]
I tested the new parameters, here are my results:
* liquid_fluidity: :heavy_check_mark:  works although even 100 has only small effect (like the docs imply)
* liquid_fluidity_smooth:  :heavy_check_mark:   couldn't figure out what it does at first but it does *something*
* liquid_sink:  :heavy_check_mark: 
* acceleration_default: seems fine
* acceleration_air: with `acceleration_air=0` you are unable to change direction after you have started jumping (hold key). However the fast mode (E) seems to bypass this. Sounds like a bug to me.
* speed_climb:  :heavy_check_mark: 
* speed_crouch: works except if you set `speed_crouch=0` you can't move at all *but* fast mode bypasses this once again.

code for convenience:
```lua
core.register_chatcommand("s", {
	func = function(name, param)
		local player = core.get_player_by_name(name)
		param = string.split(param, " ")
		player:set_physics_override({
			[param[1]] = tonumber(param[2]),
		})
	end,
})
```

**Comment by sfan5** [comment]
5.4 compatbility test: :heavy_check_mark: 

anticheat test (local server):
* speed_climb: :heavy_check_mark: 
* speed_crouch: =50 trips anticheat
* liquid_fluidity: ー (no speed increase)
* liquid_fluidity_smooth: ー (no speed increase)
* liquid_sink: :heavy_check_mark:  (no anticheat checks for Y-)
* acceleration_default / acceleration_air: ー (not related to speed)

**Comment by Wuzzy2** [comment]
Does the "action/change needed" label still apply after my changes?

**Comment by sfan5** [comment]
I haven't had time/prioritized to test this again yet.

**Comment by Wuzzy2** [comment]
No, this PR is getting too big now. I want to keep this PR as-is, unless there's a major mistake or whatever.

**Comment by Wuzzy2** [comment]
I think I have done my part, still waiting for review. Adding any other feature just further complicates things.

**Comment by appgurueu** [comment]
Shouldn't `speed_crouch` be renamed to `speed_sneak` because crouching is referred to as sneaking in all other places?

**Comment by Wuzzy2** [comment]
It's `speed_crouch` for historic reasons. All the names are based on the names of the `movement_*` settings, and it's `movement_speed_crouch`.

**Comment by sfan5** [comment]
Discussed in meeting, moved to 5.7.0 milestone since we don't have time for it now (sorry).
https://irc.minetest.net/minetest-dev/2022-07-17#i_5995098

**Comment by Zughy** [comment]
Since #12591 has been merged (to make PRs like this easier), a rebase is needed

**Comment by Wuzzy2** [comment]
Rebased! And I did a quick and dirty re-test (but not a full test). The quick and dirty test was successful.

**Comment by Wuzzy2** [comment]
Fixed a couple of bugs that snuck in. Also another rebase because of the recent acceleration changes.

@SmallJoker.

**Comment by kneekoo** [comment]
Should the effects of entity falling be treated here? Mobs and players falling make them flash red for a bit, but depending on certain in-game factors (fall damage protection, creative mode, etc), we would need to override that flash from being displayed.

**Comment by Wuzzy2** [comment]
No. This is a completely unrelated problem. The fact that player entities and non-player entities behave so differently physically, and that they share so little physics-related code, this is a big problem IMHO, but an unrelated problem.

Maybe one day entities would be able to make use of player physics directly, instead of forcing mods to awkwardly emulating them. But this will be no trivial task that will take a lot of careful thinking first (to not break anything).

This PR is exclusively about *player* physics. It doesn't even change any physics at all, it just exposes them to Lua.

**Comment by appgurueu** [comment]
> Mobs and players falling make them flash red for a bit, but depending on certain in-game factors (fall damage protection, creative mode, etc), we would need to override that flash from being displayed.

Correct me if I'm wrong, but I don't think fall damage is applied to mobs by default?

**Comment by Wuzzy2** [comment]
Entities do not have fall damage, games must code it themselves. Only players have built-in fall damage.

**Comment by Wuzzy2** [comment]
Bump.

**Comment by Zughy** [comment]
@Wuzzy2 tiny rebase needed

**Comment by Wuzzy2** [comment]
Rebase done!

**Comment by IniKiwi** [comment]
best feature ever!

**Comment by jamesalexanderstevenson** [comment]
Oh shit.

**Comment by Nukiloco** [comment]
What's the progress on this so far?

**Comment by appgurueu** [comment]
It seems to be ready for review, but won't make it into 5.7, as the feature freeze has already started.

@Zughy please remove the "rebase needed" label; there are no conflicts with the base branch anymore.

**Comment by Wuzzy2** [comment]
Yes, it is waiting for review. Unfortunately, it has missed the 5.7.0 milestone. :-(
Like it has missed the 5.6.0 milestone and 5.5.0 milestones before. :-(((
(yes, this PR is old, lol)

And sadly, I am not allowed to update labels.

**Comment by Wuzzy2** [comment]
Please someone test this PR (no matter if a rebase is neccessary or not). There are still zero approvals. I really want this to be in 5.8.0, esepecially since it has already missed inclusion in 3 Minetest releases.

**Comment by fluxionary** [comment]
> Please someone test this PR (no matter if a rebase is neccessary or not). There are still zero approvals. I really want this to be in 5.8.0, esepecially since it has already missed inclusion in 3 Minetest releases.

is this something that a core dev needs to test, or would a mod developer's help be useful? i'd love to see these features myself.

**Comment by Zughy** [comment]
@Wuzzy2 please rebase docs (see #13438)

**Comment by Wuzzy2** [comment]
Done.

File: src/client/clientenvironment.cpp
Comment (SmallJoker) [inline src/client/clientenvironment.cpp:228]: Possible division by zero.

File: doc/lua_api.md
Comment (SmallJoker) [inline doc/lua_api.md:7462]: Why? Because of the division by zero below? If there is a limit, it should be enforced in the code too (e.g. `MYMAX`).

**Review by SmallJoker — COMMENTED** [review]
Did a few tests including negative values: https://github.com/SmallJoker/minetest_lab/blob/master/init_137_pysics_override_11465.lua

Last few comments before approval.

**Comment by Wuzzy2** [comment]
I added `MYMAX` in the code as requested.
The reason I want to enforce a lower limit of 1 is because at very low values, the movement physics seems to kinda break a little. This was a problem before btw (if you modified the `movement_*` setting) so not a regression.

So better safe than sorry. Maybe later the liquid physics can be improved so that low numbers will work properly as well but this needs careful experimentation and stuff. Remember, this PR does NOT want to change the physics in any way but merely expose physics values to the API.

And yes, if liquid\_fluidity is 0, that's a division by zero and if you enter water, you teleport to -NaN, -NaN, -NaN. This was a bug but with the new limit enforcements, the division by 0 is gone now.

**Review by SmallJoker — APPROVED** [review]
Works.

**Comment by sfan5** [comment]
@SmallJoker if the bugs I mentioned have been fixed then feel free to merge.

File: src/server/player_sao.cpp
Comment (grorp) [inline src/server/player_sao.cpp:645]: This line is duplicated. I think only the one after `player_max_jump *= 2.0;` should stay.

See also: https://github.com/minetest/minetest/pull/11465#discussion_r760271880

File: src/client/content_cao.cpp
Comment (grorp) [inline src/client/content_cao.cpp:1792]:
```suggestion
		// fallback for new overrides (since 5.8.0)
```

File: src/server/player_sao.cpp
Comment (grorp) [inline src/server/player_sao.cpp:328]:
```suggestion
	// new physics overrides since 5.8.0
```

**Review by grorp — CHANGES_REQUESTED** [review]
My three comments are just on minor things. I'll test this soon.

**Comment by grorp** [comment]
Tested. Everything works, except negative `liquid_fluidity_smooth` values. These cause buggy behavior, including teleportation to NaN. See the end of this video: https://youtu.be/rirGBG5-QaQ. I suggest disallowing negative values for `liquid_fluidity_smooth`.

I haven't found any other issues with negative values or zero values, not even with `acceleration_default` and `acceleration_air`.

Regarding some previous comments by sfan5 and Wuzzy2:

> acceleration_air: with acceleration_air=0 you are unable to change direction after you have started jumping (hold key). However the fast mode (E) seems to bypass this. Sounds like a bug to me.
> https://github.com/minetest/minetest/pull/11465#issuecomment-983756913

I would expect `movement_acceleration_fast` to take precedence over both `movement_acceleration_default` and `movement_acceleration_air`, so this seems correct to me.

> speed_crouch: works except if you set speed_crouch=0 you can't move at all but fast mode bypasses this once again.
> https://github.com/minetest/minetest/pull/11465#issuecomment-983756913

I think that's expected too, at least `movement_speed_fast` already overrides sneak speed on the `master` branch if you sneak and "sprint" at the same time.

> Should I try to fix the Fast Mode behavior anyway? Because I think you're right, Fast Mode shouldn't be able to bypass this.
> https://github.com/minetest/minetest/pull/11465#issuecomment-1005992854

If you're referring to the fact that fast speed takes precedence when you sneak and "sprint" at the same time:  I actually don't think this behavior is bad, it allows you to use fast mode with fall-off protection.

> Finally, I think I should actually just add a fast speed override as well. Do you agree?
> https://github.com/minetest/minetest/pull/11465#issuecomment-1005992854

I agree that it would probably be good to allow overriding `movement_speed_fast` and `movement_acceleration_fast` as well, but obviously not in this PR anymore.

File: src/client/clientenvironment.cpp
Comment (grorp) [inline src/client/clientenvironment.cpp:225]:
```suggestion
				fluidity_smooth *= lplayer->physics_override.liquid_fluidity_smooth;
				fluidity_smooth = MYMAX(0.0f, fluidity_smooth);
```

See https://github.com/minetest/minetest/pull/11465#issuecomment-1703853842.

**Comment by grorp** [comment]
I'll approve once my four comments are resolved.

**Comment by Wuzzy2** [comment]
@grorp. Done. I agree to not add `fast` support into this PR. Only bugfixes and "PR maintenance" from this point on. This PR has been waiting for 2 years already.

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:7686]:
```suggestion
        * `speed`: multiplier to default movement speed and acceleration values (default: `1`)
```

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:7705]:
```suggestion
          on ground or when climbing (default: `1`)
            * Note: The actual acceleration is the product of `speed` and `acceleration_default`
```

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:7707]:
```suggestion
          when jumping or falling (default: `1`)
            * Note: The actual acceleration is the product of `speed` and `acceleration_air`
```

**Review by grorp — CHANGES_REQUESTED** [review]
I'm sorry, but I found another thing: It isn't documented that `speed` also affects acceleration values. See here:

https://github.com/minetest/minetest/blob/c1c05df00e645862fe1074d93661491629383b3b/src/client/localplayer.cpp#L694-L697

See my change suggestions below.

**Comment by Wuzzy2** [comment]
@grorp: Done.

File: src/server/player_sao.cpp
Comment (grorp) [inline src/server/player_sao.cpp:328]:
```suggestion
	// new physics overrides since 5.8.0
```

The typo is still there.

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:7697]:
```suggestion
            * Warning: Values below 1 are currently unsupported.
```

Otherwise, the line break won't be visible in the rendered version.

**Review by grorp — CHANGES_REQUESTED** [review]
...

**Comment by appgurueu** [comment]
@Wuzzy2 consider allowing edits by maintainers in the future in order to avoid such round trips for trivial fixes.

**Comment by Wuzzy2** [comment]
But ... but ... edits by maintainers are already enabled? :-(

**Comment by Wuzzy2** [comment]
@grorp: Done.

**Review by grorp — APPROVED** [review]
LGTM.

**Comment by grorp** [comment]
> consider allowing edits by maintainers in the future in order to avoid such round trips for trivial fixes.

> But ... but ... edits by maintainers are already enabled? :-(

I had forgotten about that feature.

**Comment by appgurueu** [comment]
> But ... but ... edits by maintainers are already enabled? :-(

Heh, it seems GitHub shows that ticked checkbox to maintainers or PR authors, but not to everyone else. So on the MTG repo or on my own PRs I'm used to seeing it, so I assumed its absence here (paired with the fact that core devs didn't take advantage of it) meant it hadn't been enabled. My bad.

**Comment by Wuzzy2** [comment]
Hooray! Finally merged. :-)

I have posted a follow-up PR for Fast Mode: #13815

**Pull request merged**
